src/OpenDebugAD7/AD7DebugSession.cs

```diff
@@ -745,6 +746,12 @@ protected override void HandleInitializeRequestAsync(IRequestResponder<Initializ
 
         protected override void HandleLaunchRequestAsync(IRequestResponder<LaunchArguments> responder)
         {
+            if (clientId == "visualstudio" && !responder.Arguments.ConfigurationProperties.GetValueAsBool("externalConsole").GetValueOrDefault(false))
+            {
+                responder.SetError(new ProtocolException("Integrated terminal is not supported in Visual Studio. To fix, set external console to true."));
```

Comment:
>"Integrated terminal is not supported in Visual Studio. To fix, set external console to true." [](start = 57, length = 94)

If we decide we want to fail, this should be localized.

Reply:
Correct me if I am wrong, but the main difference between the two options - one to fail and the other to ignore useExternalConsole - is the user awareness of this issue. Perhaps it would be better from the customer's perspective if this error dialog does not appear at all.

What are your thoughts, @gregg-miskelly  and @WardenGnaw?

Reply:
I sent you an email about this last night, but I would suggest ignoring `useExternalConsole` if the host IDE is Visual Studio. Three main reasons for this:

1. If we were in a world where one is trying to share launch.json content between VS and VS Code, it seems better for VS to ignore the option rather than failing.
2. The fact that the default value of this option is `false` means that we would otherwise require it to be specified, which really seems wrong. We could definitely fix this instead (turn this into a `bool?` throughout the API), but the effort doesn't seem worth it.
3. This is violates layering -- OpenDebugAD7 now has hard coded knowledge of an MIEngine-specific option

Reply:
I just saw the email. Thank you for bringing that up. One question - by "if the host IDE is Visual Studio", you are referring to both when ClientId = VisualStudio and ClientId = LiveshareServerHost, correct?

Reply:
Correct.

Reply:
Thank you!

Reply:
I have spent quite some time adjusting my approach to ignore useExternalConsole when IsClientVS is true. However, the below error dialog continues to appear when externalConsole is set to false in the test launch.json file:
![image](https://user-images.githubusercontent.com/60245970/101226873-b9588c80-364a-11eb-9b52-f9a3375be99c.png)

Here is the corresponding Debug Adapter Host Log output:
```
 1> [DebugAdapter] --> C (launch-2): {"type":"request","command":"launch","arguments":{"name":"g++.exe - Build and debug active file","type":"cppdbg","request":"launch","program":"C:\\Users\\gabchen\\Downloads\\projects\\helloworld\\helloworld.exe","args":[],"stopAtEntry":false,"cwd":"C:\\Users\\gabchen\\Downloads\\projects\\helloworld","environment":[],"externalConsole":false,"MIMode":"gdb","miDebuggerPath":"D:\\Program Files (x86)\\mingw-w64\\i686-8.1.0-posix-dwarf-rt_v6-rev0\\mingw32\\bin\\gdb.exe","setupCommands":[{"description":"Enable pretty-printing for gdb","text":"-enable-pretty-printing","ignoreFailures":true}],"preLaunchTask":"C/C++: g++.exe build active file"},"seq":2}
 1> [DebugAdapter] <--   E (output): {"type":"event","event":"output","body":{"category":"stderr","output":"Internal file \"c:\\program files (x86)\\microsoft visual studio\\2019\\main\\common7\\ide\\commonextensions\\microsoft\\mdd\\debugger\\WindowsDebugLauncher.exe\" could not be found.\r\n\r\n"},"seq":2}
 1> [DebugAdapter] <--   R (launch-2) [246 ms]: {"type":"response","request_seq":2,"success":false,"command":"launch","message":"Unable to start debugging. Unable to establish a connection to GDB. The following message was written to stderr:\r\n\r\nInternal file \"c:\\program files (x86)\\microsoft visual studio\\2019\\main\\common7\\ide\\commonextensions\\microsoft\\mdd\\debugger\\WindowsDebugLauncher.exe\" could not be found.\r\n\r\nSee Output Window for details.","body":{"error":{"id":1005,"format":"Unable to start debugging. Unable to establish a connection to GDB. The following message was written to stderr:\r\n\r\nInternal file \"c:\\program files (x86)\\microsoft visual studio\\2019\\main\\common7\\ide\\commonextensions\\microsoft\\mdd\\debugger\\WindowsDebugLauncher.exe\" could not be found.\r\n\r\nSee Output Window for details."}},"seq":3}
 1> ERROR: Unexpected error

AggregateException: One or more errors occurred.

Aggregate exception: 
    DebugAdapterLaunchException: Failed to launch debug adapter.  Additional information may be available in the output window.

    Failure Location: LaunchRequest
    Inner Exception: 
        ProtocolException: Unable to start debugging. Unable to establish a connection to GDB. The following message was written to stderr:
        
        Internal file "c:\program files (x86)\microsoft visual studio\2019\main\common7\ide\commonextensions\microsoft\mdd\debugger\WindowsDebugLauncher.exe" could not be found.
        
        See Output Window for details.
Inner Exception: 
    DebugAdapterLaunchException: Failed to launch debug adapter.  Additional information may be available in the output window.

Microsoft.VisualStudio.Debugger.VSCodeDebuggerHost.Engine.Implementation.DebuggedProcess.<StartDebugAdapter>b__114_3(Exception ex)
Microsoft.VisualStudio.Debugger.VSCodeDebuggerHost.Utilities.TaskExtensions.<>c__DisplayClass11_0`1.<Catch>b__0(TException ex)
Microsoft.VisualStudio.Debugger.VSCodeDebuggerHost.Utilities.TaskExtensions.<>c__DisplayClass10_0`1.<Catch>b__0(AggregateException ex)

    Failure Location: LaunchRequest
    Inner Exception: 
        ProtocolException: Unable to start debugging. Unable to establish a connection to GDB. The following message was written to stderr:
        
        Internal file "c:\program files (x86)\microsoft visual studio\2019\main\common7\ide\commonextensions\microsoft\mdd\debugger\WindowsDebugLauncher.exe" could not be found.
        
        See Output Window for details.

 1> ERROR: One or more errors occurred.

Failed to launch debug adapter.  Additional information may be available in the output window.

Unable to start debugging. Unable to establish a connection to GDB. The following message was written to stderr:

Internal file "c:\program files (x86)\microsoft visual studio\2019\main\common7\ide\commonextensions\microsoft\mdd\debugger\WindowsDebugLauncher.exe" could not be found.

See Output Window for details.
 1> Debug adapter process exited.
```

Reply:
Note: the externalConsole = true in launch.json scenario works as expected; the modules show in the Modules window.

Some extra information that may be helpful:
- clientId = "visualstudio"
- m_clientId = VisualStudio
- IsClientVS = true

Reply:
This is because we don't install WindowsDebugLauncher.exe with VS.

@WardenGnaw Is the scenario in this bug (launching a mingw program) actually a Codespaces scenario? Or is this really just a test scenario?

If yes - the right way to fix this is to add WindowsDebugLauncher.exe to setup.
If no - the right way to fix this is to stop calling `RegisterRunInTerminalCallback` so that (IsRunInTerminalAvailable)(https://github.com/gregg-miskelly/MIEngine/blob/c9dd86298463a60df043148e3cab49f4ad08994b/src/DebugEngineHost.VSCode/HostRunInTerminal.cs#L19-L22) will return false.

Reply:
BTW: If we do need to support mingw with Codespaces, then [this code](https://github.com/gregg-miskelly/MIEngine/blob/c9dd86298463a60df043148e3cab49f4ad08994b/src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs#L766-L772) is wrong if `UseExternalConsole` can ever return `true` -- all Codespaces launches of mingw apps need to use RunInTerminal. So we might need a new `HostRunInTerminal` method that we could use to tell MIEngine not to use `-gdb-set new-console on`.

Reply:
I'm actually not sure. @andrewcrawley Added OpenDebugAD7 to Codespaces initially, was there not a need for adding this binary?

We ran into this issue because CTI created this `launch.json` with `externalConsole: false` and I'm assuming they are also doing the local launch scenario with DebugAdapterHost.Launch instead of testing in a CodeSpace.

Reply:
Should be good to go now. 

Here is what the Modules window looks like when externalConsole is false:
![image](https://user-images.githubusercontent.com/60245970/101531997-c75b2580-3948-11eb-8e60-a27505b2921b.png)

And here is what the Modules window looks like when externalConsole is false:
![image](https://user-images.githubusercontent.com/60245970/101532035-d2ae5100-3948-11eb-855d-170c3f1c0bf9.png)